components/use-invoice.js

```diff
@@ -16,14 +16,21 @@ export default function useInvoice () {
       throw error
     }
 
-    const { cancelled, cancelledAt, actionError, actionState, expiresAt } = data.invoice
+    const { cancelled, cancelledAt, actionError, actionState, expiresAt, isHeld } = data.invoice
 
     const expired = cancelledAt && new Date(expiresAt) < new Date(cancelledAt)
     if (expired) {
       throw new InvoiceExpiredError(data.invoice)
     }
 
-    if (cancelled || actionError) {
+    const failed = cancelled || actionError
+
+    // failed forwards might already have been finalized (actionState === 'FAILED') so we also check for failed held payments
+    if (actionState === 'FAILED_FORWARD' || (failed && isHeld)) {
+      throw new WalletReceiverError(data.invoice)
+    }
```

Comment:
It is unfortunate that invoice polling can confusingly also return a wallet error now but I think that's not too bad for the sake of keeping this simple.

Alternative would be to manually query the invoice on payment errors to check for failed forwards but I think the flow is easier to understand if the receiver error is thrown here.